src/analyzer.ts

```diff
@@ -8,762 +8,139 @@
-        if (metadata.elements) {
-          metadata.elements.forEach(function(element) {
-            attachDomModule(parsedHtml, element);
-          });
-        }
-        return metadata;
-      });
-    });
-  };
+  async analyzeSource(type: string, contents: string, url: string)
+      : Promise<DocumentDescriptor>  {
+    let document = this.parse(type, contents, url);
+    return this.analyzeDocument(document);
+  }
 
+  async analyzeDocument(document: Document<any, any>)
+      : Promise<DocumentDescriptor> {
+    let entities = await this.getEntities(document);
 
-  _inlineStyles(ast:dom5.Node, href: string) {
-    var cssLinks = dom5.queryAll(ast, polymerExternalStyle);
-    cssLinks.forEach((link) => {
-      var linkHref = dom5.getAttribute(link, 'href');
-      var uri = url.resolve(href, linkHref);
-      var content = this._content[uri];
-      var style = dom5.constructors.element('style');
-      dom5.setTextContent(style, '\n' + content + '\n');
-      dom5.replace(link, style);
-    });
-    return cssLinks.length > 0;
-  };
+    // TODO(justinfagnani): Load ImportDescriptors
 
-  _inlineScripts(ast: dom5.Node, href: string) {
-    var scripts = dom5.queryAll(ast, externalScript);
-    scripts.forEach((script) => {
-      var scriptHref = dom5.getAttribute(script, 'src');
-      var uri = url.resolve(href, scriptHref);
-      var content = this._content[uri];
-      var inlined = dom5.constructors.element('script');
-      dom5.setTextContent(inlined, '\n' + content + '\n');
-      dom5.replace(script, inlined);
-    });
-    return scripts.length > 0;
-  };
+    return new DocumentDescriptor(document, entities);
+  }
 
-  _inlineImports(ast:dom5.Node, href:string, loaded:{[href:string]:boolean}) {
-    var imports = dom5.queryAll(ast, isHtmlImportNode);
-    imports.forEach((htmlImport) => {
-      var importHref = dom5.getAttribute(htmlImport, 'href');
-      var uri = url.resolve(href, importHref);
-      if (loaded[uri]) {
-        dom5.remove(htmlImport);
-        return;
-      }
-      var content = this.getLoadedAst(uri, loaded);
-      dom5.replace(htmlImport, content);
-    });
-    return imports.length > 0;
-  };
 
   /**
-   * Returns a promise resolving to a form of the AST with all links replaced
-   * with the document they link to. .css and .script files become &lt;style&gt; and
-   * &lt;script&gt;, respectively.
-   *
-   * The elements in the loaded document are unmodified from their original
-   * documents.
-   *
-   * @param {string} href The document to load.
-   * @param {Object.<string,boolean>=} loaded An object keyed by already loaded documents.
-   * @return {Promise.<DocumentAST>}
+   * Loads and parses a single file, deduplicating any requrests for the same
+   * URL.
    */
-  getLoadedAst(href:string, loaded:{[href:string]:boolean}) {
-    if (!loaded) {
-      loaded = {};
+  async load(url: string): Promise<Document<any, any>> {
+    // TODO(justinfagnani): normalize url
```

Comment:
For normalization to be effective, the entire document tree needs to get resolved against the same Base.